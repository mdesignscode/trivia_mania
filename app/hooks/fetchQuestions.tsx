"use client";

import { GlobalContext } from "@/context/globalContext";
import { useQuery } from "@tanstack/react-query";
import axios from "axios";
import { useContext, useEffect, useState } from "react";
import useGameStore from "../game/components/store";

interface IFetchQuestionsProps {
  categories: string;
  difficulty: string;
}

export default function useFetchQuestions({
  difficulty,
  categories,
}: IFetchQuestionsProps) {
  const baseUrl = process.env.NEXT_PUBLIC_API_BASE_URL;
  const url = baseUrl + "questions/play";
  const { setQuestions } = useGameStore();
  const {
      storageIsAvailable,
      newFilters,
      triviaUser: user,
    } = useContext(GlobalContext),
    [enabled, setEnabled] = useState(false);

  // check for previous set filters
  useEffect(() => {
    // fetch new questions for users without local storage
    if (!storageIsAvailable) {
      setEnabled(true);
      return;
    }

    // check for first time render on client
    const [localDifficulty, localCategories] = [
      localStorage.getItem("difficulty"),
      localStorage.getItem("categories"),
    ];

    if (!localDifficulty && !localCategories) {
      // fetch new questions
      setEnabled(true);

      // set new filters in local storage
      localStorage.setItem("difficulty", difficulty);
      localStorage.setItem("categories", categories);

      return;
    }

    // fetch new questions if user answered questions updates
    // and game page refreshes
    const invalidateQuestions = localStorage.getItem("fetchNewQuestions"),
      gamePageMounted = localStorage.getItem("gamePageMounted");

    if (invalidateQuestions === "true" && gamePageMounted === "false") {
      setEnabled(true);
      return;
    }

    // check if new filters match
    if (!newFilters) return;

    // fetch new questions
    setEnabled(true);
  }, [categories, difficulty, newFilters, storageIsAvailable]);

  useQuery<TQuestion[]>({
    queryKey: [`fetchQuestions`, difficulty, categories],
    queryFn: async () => {
      try {
        const { data } = await axios.post(url, {
          categories,
          difficulty,
          user,
        });
        console.log(data)

        if (storageIsAvailable)
          localStorage.setItem("questions", JSON.stringify(data));

        setQuestions(data);

        return data;
      } catch (error) {
        console.log(error);
      }
    },
    initialData: () => {
      const prevData = storageIsAvailable
        ? (JSON.parse(localStorage.getItem("questions") || "[]") as TQuestion[])
        : ([] as TQuestion[]);
      setQuestions(prevData);
      setQuestions(prevData);
      return prevData;
    },
    enabled,
  });
}
