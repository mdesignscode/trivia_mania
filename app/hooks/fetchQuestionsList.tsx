/* Performs questions fetching */
"use client";
import { GlobalContext } from "@/app/context/globalContext";
import { IPlayRequest } from "@/models/customRequests";
import { IQuestion } from "@/models/interfaces";
import Question from "@/models/question";
import {
  CATEGORIES,
  CURRENT_INDEX,
  DIFFICULTY,
  NEW_PARAMS,
  POOL_INDEX,
  QUESTIONS_LIST,
  QUESTIONS_POOL,
  QUESTION_ANSWERED,
  clearQuestionData,
} from "@/utils/localStorage_utils";
import { useQuery } from "@tanstack/react-query";
import axios from "axios";
import { useCallback, useContext, useEffect, useRef, useState } from "react";

interface IFetchQuestions {
  questions: IQuestion[];
  nextQuestionsSet: () => void;
  questionsLength: number;
  questionIndex: number;
  incrementIndex: () => void;
  questionsPoolReady: boolean;
  poolIndex: number;
  setQuestionsPool: (state: IQuestion[]) => void;
}

export default function useFetchQuestionsList(): IFetchQuestions {
  const [_questionIndex, _setQuestionIndex] = useState(1);
  const questionIndex = useRef(_questionIndex);
  function setQuestionIndex(state: number) {
    questionIndex.current = state;
    _setQuestionIndex(state);
  }
  const [_questionsPoolReady, _setQuestionsPoolReady] = useState(false);
  const [localStorageReady, setLocalStorageReady] = useState(false);
  const questionsPoolReady = useRef(_questionsPoolReady);
  function setQuestionsPoolReady(state: boolean) {
    questionsPoolReady.current = state;
    _setQuestionsPoolReady(state);
  }

  // render 5 questions at a time
  const [_questionsPool, _setQuestionsPool] = useState<IQuestion[]>([]);
  const questionsPool = useRef(_questionsPool);
  function setQuestionsPool(state: IQuestion[]) {
    questionsPool.current = state;
    _setQuestionsPool(state);
  }
  const [_poolIndex, _setPoolIndex] = useState(0);
  const poolIndex = useRef(_poolIndex);
  const setPoolIndex = useCallback((index: number) => {
    poolIndex.current = index;
    _setPoolIndex(index);
  }, []);
  const [questionsLength, setQuestionsLength] = useState(0);

  const [shouldFetchQuestions, setShouldFetchQuestions] = useState(false);
  const [fetchQuestions, setFetchQuestions] = useState(true);

  const {
    triviaUser,
    playFilters: { difficulty, categories },
    storageIsAvailable,
    setNewFilters,
    newFilters,
  } = useContext(GlobalContext);

  // query function
  async function getQuestions() {
    const baseUrl = process.env.NEXT_PUBLIC_API_BASE_URL;
    const url = `${baseUrl}questions/play`;

    const { data } = await axios.post(url, {
      difficulty,
      categories: categories.split(",").filter(Boolean),
      userId: triviaUser?.id,
    } as IPlayRequest);

    const questions = shuffleArray(data);

    return questions;
  }

  // fetch list of questions
  const { data, isFetched } = useQuery({
    queryKey: ["play", triviaUser, difficulty, categories],
    queryFn: getQuestions,
    initialData: [],
    enabled: shouldFetchQuestions,
  });

  // create return object
  const fetchingUtils: IFetchQuestions = {
    questions: questionsPool.current,
    questionsLength,
    nextQuestionsSet,
    questionIndex: questionIndex.current,
    incrementIndex,
    questionsPoolReady: questionsPoolReady.current,
    poolIndex: poolIndex.current,
    setQuestionsPool,
  };

  // create initial questions pool
  const createQuestionPool = useCallback(() => {
    if (localStorageReady) {
      const localQuestionsPool = localStorage.getItem(QUESTIONS_POOL),
        localQuestionsList: Question[] = JSON.parse(
          localStorage.getItem(QUESTIONS_LIST) || "[]"
        ),
        localIndex = parseInt(localStorage.getItem(CURRENT_INDEX) || "1"),
        localPoolIndex = parseInt(localStorage.getItem(POOL_INDEX) || "0");

      if (localQuestionsPool) {
        setQuestionsPool(JSON.parse(localQuestionsPool));
        setPoolIndex(localPoolIndex);
      } else {
        const newIndex = calculateNewIndex(0, localQuestionsList.length),
          newPool = localQuestionsList.slice(0, newIndex);

        setQuestionsPool(newPool);
        localStorage.setItem(QUESTIONS_POOL, JSON.stringify(newPool));
      }

      localStorage.setItem(CURRENT_INDEX, questionIndex.current.toString());
      localStorage.setItem(POOL_INDEX, localPoolIndex.toString());

      setQuestionsLength(localQuestionsList.length);
      setQuestionIndex(localIndex);
      setLocalStorageReady(false);
      setQuestionsPoolReady(true);
    }
  }, [localStorageReady, setPoolIndex]);

  // fetch questions from api and store in local storage
  const fetchNewQuestions = useCallback(() => {
    if (fetchQuestions) {
      if (newFilters) {
        setShouldFetchQuestions(true);
        if (isFetched) {
          localStorage.setItem(QUESTIONS_LIST, JSON.stringify(data));
          localStorage.setItem(CATEGORIES, categories);
          localStorage.setItem(DIFFICULTY, difficulty);
          setLocalStorageReady(true);
          setFetchQuestions(false);
          setNewFilters(false);
        }
      } else {
        setLocalStorageReady(true);
        setFetchQuestions(false);
      }
    }
  }, [categories, data, difficulty, fetchQuestions, isFetched, newFilters, setNewFilters]);

  // set questions pool
  useEffect(() => {
    if (storageIsAvailable) {
      fetchNewQuestions();
      createQuestionPool();
    } else {
      if (fetchQuestions) {
        setShouldFetchQuestions(true);
        if (isFetched) {
          setQuestionsLength(data.length);
          setQuestionsPool(data.slice(0, 5));
          setFetchQuestions(false);
        }
      }
    }
  }, [
    createQuestionPool,
    data,
    fetchNewQuestions,
    fetchQuestions,
    isFetched,
    storageIsAvailable,
  ]);

  // fetch next set of questions from list
  function nextQuestionsSet() {
    setQuestionsPoolReady(false);

    const newIndex = calculateNewIndex(poolIndex.current, questionsLength);
    setPoolIndex(newIndex);

    const endIndex = calculateEndIndex(newIndex, questionsLength);

    if (storageIsAvailable) {
      // get questions list local storage
      const localQuestionsList = JSON.parse(
        localStorage.getItem(QUESTIONS_LIST) || "[]"
      ) as IQuestion[];

      incrementIndex();

      const newPool = localQuestionsList.slice(newIndex, endIndex);

      // set next set of questions
      setQuestionsPool(newPool);

      // increment pool index in memory and locally
      localStorage.setItem(POOL_INDEX, newIndex.toString());
      localStorage.setItem(QUESTIONS_POOL, JSON.stringify(newPool));
    } else {
      incrementIndex();
      setQuestionsPool((data as IQuestion[]).slice(newIndex, endIndex));
    }
    setQuestionsPoolReady(true);
  }

  function incrementIndex() {
    setQuestionIndex(questionIndex.current + 1);

    if (storageIsAvailable) {
      localStorage.setItem(QUESTION_ANSWERED, "false");
      localStorage.setItem(CURRENT_INDEX, questionIndex.current.toString());
    }
  }

  return fetchingUtils;
}

// shuffle questions
function shuffleArray(array: Array<IQuestion>): Array<IQuestion> {
  for (let i = array.length - 1; i > 0; i--) {
    const j = Math.floor(Math.random() * (i + 1));
    [array[i], array[j]] = [array[j], array[i]]; // Swap elements
  }
  return array;
}

export function calculateNewIndex(poolIndex: number, questionsLength: number) {
  return poolIndex + 5 > questionsLength
    ? questionsLength - poolIndex + poolIndex // last set of questions
    : poolIndex + 5;
}

export function calculateEndIndex(newIndex: number, questionsLength: number) {
  return newIndex + 5 > questionsLength ? questionsLength : newIndex + 5;
}
