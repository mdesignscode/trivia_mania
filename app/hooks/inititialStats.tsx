"use client";

import { useUser } from "@clerk/nextjs";
import { GlobalContext } from "app/store";
import axios from "axios";
import { useContext, useEffect, useState } from "react";

type TStat = Record<string, number>;
export interface IInitialStats {
  difficultyStats: TStat;
  categoryStats: TStat;
  categoriesLoading: boolean;
  difficultiesLoading: boolean;
  previousDifficulty: string;
}

export default function useInitialStats(): IInitialStats {
  const [initialStats, setInitialStats] = useState<IInitialStats>({
    difficultyStats: {},
    categoryStats: {},
    categoriesLoading: true,
    difficultiesLoading: true,
    previousDifficulty: "",
  });
  const { user, isLoaded, isSignedIn } = useUser();
  const { storageIsAvailable } = useContext(GlobalContext);

  // request body
  let reqBody = {
    difficulty: "",
    userId: "",
  };

  async function fetchInitialQuestionStats() {
    // wait for clerk to finish loading
    if (isLoaded) {
      // if user is available
      if (isSignedIn) {
        // add user id to request body
        reqBody.userId = user.id;

        // check for previous difficulty set
        if (storageIsAvailable) {
          const prevDiff = localStorage.getItem("difficulty");
          if (prevDiff) {
            // add difficulty to request body
            reqBody.difficulty = prevDiff;
          }
        }
      }

      try {
        // prepare url for question stats
        const baseUrl = process.env.NEXT_PUBLIC_API_BASE_URL;
        const url = baseUrl + "questions/stats";

        // fetch initial questions stats
        // get categories based on difficulty
        const { data } = await axios.post(url, reqBody);
        // get difficulty stats, which is not included above
        const { data: diffStats } = await axios.post(url, { userId: reqBody.userId });

        const difficultyStats: TStat = {};
        const categoryStats: TStat = {};

        for (const key in data) {
          // set categories stats
          categoryStats[key] = data[key];
        }

        for (const key in diffStats) {
          // set difficulty stats
          if (["easy", "medium", "hard", "all difficulties"].includes(key)) {
            difficultyStats[key] = diffStats[key];
          }
        }

        setInitialStats({
          difficultyStats,
          categoryStats,
          categoriesLoading: false,
          previousDifficulty: reqBody.difficulty,
          difficultiesLoading: false,
        });
      } catch (error) {
        console.log(error);
      }
    }
  }

  useEffect(() => {
    fetchInitialQuestionStats();
  }, [isLoaded, isSignedIn, storageIsAvailable]);

  return initialStats;
}
