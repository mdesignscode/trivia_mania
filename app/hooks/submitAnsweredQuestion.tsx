"use client";

import { useQuery } from "@tanstack/react-query";
import axios from "axios";
import { useContext, useState } from "react";
import { GlobalContext } from "../context/globalContext";

interface ISubmitAnsweredQuestionProps {
  answeredCorrect: boolean;
  question: TQuestion;
}

interface IUpdateStatsBody {
  user: NonNullable<TUser>;
  answeredCorrect: boolean;
  question: TQuestion;
}

export default function useSubmitAnsweredQuestion({
  answeredCorrect,
  question,
}: ISubmitAnsweredQuestionProps) {
  const { triviaUser } = useContext(GlobalContext);
  const [shouldSubmit, setshouldSubmit] = useState(false);

  const baseUrl = process.env.NEXT_PUBLIC_API_BASE_URL;
  const url = baseUrl + "users/updateStats";

  const {} = useQuery({
    queryKey: [`submitQuestion-${triviaUser?.id}-${question.id}`],
    queryFn: async () => {
      const postBody: IUpdateStatsBody = {
        user: triviaUser as NonNullable<TUser>,
        question,
        answeredCorrect,
      };
      try {
        const { data } = await axios.post(url, postBody);

        return data;
      } catch (error) {
        console.log(error);
      }
    },
    enabled: shouldSubmit,
    onSuccess: () => setshouldSubmit(false)
  });

  return { setshouldSubmit };
}
