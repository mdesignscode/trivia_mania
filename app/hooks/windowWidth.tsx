import { useEffect, useState } from "react";

export default function useWindowWidth() {
  const [windowWidth, detectWidth] = useState(window.innerWidth);
  const detectSize = () => {
    detectWidth(window.innerWidth);
  };
  useEffect(() => {
    window.addEventListener('resize', detectSize);
    return () => {
      window.removeEventListener('resize', detectSize);
    };
  }, [windowWidth]);

  return windowWidth < 769 // mobile
}
