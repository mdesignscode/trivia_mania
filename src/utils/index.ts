import { browser } from '$app/environment';

export function decodeHTMLEntities(text: string) {
        if (browser) {
                const element = document.createElement("div");
                element.innerHTML = text;
                return element.textContent || "";
        }
}

export const colorMap: { [key: string]: string } = {
        easy: "green",
        medium: "gold",
        hard: "red",
};

export function capitalize(str: string): string {
	if (!str) return '';
	return str.charAt(0).toUpperCase() + str.slice(1).toLowerCase();
}

