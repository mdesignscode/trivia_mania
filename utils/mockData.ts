import { CategoryStat, DifficultyStat, IUserStats } from "@/models/interfaces";
import { IQuestion } from "@/models/interfaces";

type TStat = Record<string, number>

export const mockInitialProgress: IUserStats = {
  total: {
    correctAnswered: 1,
    answered: 1,
  },
  easy: {
    answered: 1,
    correctAnswered: 1,
  } as DifficultyStat,
  Science: {
    easy: {
      answered: 1,
      correctAnswered: 1,
    } as DifficultyStat,
  } as CategoryStat,
};

// create mock questions
export const mockQuestions: Record<string, IQuestion> = {
  "63805aab-b589-4256-9493-769eecc23966": {
    "category": "Entertainment: Film",
    "answers": [
      "Batman v Superman: Superapocalypse",
      "Batman v Superman: Black of Knight",
      "Batman v Superman: Dawn of Justice",
      "Batman v Superman: Knightfall"
    ],
    "correctAnswer": "Batman v Superman: Dawn of Justice",
    "id": "63805aab-b589-4256-9493-769eecc23966",
    "question": "Which movie released in 2016 features Superman and Batman fighting?",
    "difficulty": "medium"
  },
  "36240604-80aa-4ddb-bb87-5ce345bfd379": {
    "category": "Entertainment: Film",
    "answers": ["R2-D2", "AA-A", "BB-8", "BB-3"],
    "correctAnswer": "BB-8",
    "id": "36240604-80aa-4ddb-bb87-5ce345bfd379",
    "question": "What is the orange and white bot&#039;s name in &quot;Star Wars: The Force Awakens&quot;?",
    "difficulty": "easy"
  },
  "12c4b40f-7fa9-405d-81de-b64c54a0caea": {
    "category": "Entertainment: Music",
    "answers": [
      "James Blunt - &quot;1973&quot;",
      "Prince - &quot;1999&quot;",
      "The Smashing Pumpkins - &quot;1979&quot;",
      "David Bowie - &quot;1984&quot;"
    ],
    "correctAnswer": "The Smashing Pumpkins - &quot;1979&quot;",
    "id": "12c4b40f-7fa9-405d-81de-b64c54a0caea",
    "question": "Which of these songs was released in 1996?",
    "difficulty": "hard"
  },
  "b2d0c4b4-ac68-491c-a561-60848ea34fb6": {
    "category": "Entertainment: Music",
    "answers": [
      "Gimmie! Gimmie! Gimme! (A Man After Midnight)",
      "Night Fever",
      "The Chain",
      "Staying Alive"
    ],
    "correctAnswer": "Gimmie! Gimmie! Gimme! (A Man After Midnight)",
    "id": "b2d0c4b4-ac68-491c-a561-60848ea34fb6",
    "question": "Madonna&#039;s song &quot;Hung Up&quot; includes a piece from which popular 70s song?",
    "difficulty": "medium"
  },
  "83cea789-1c6d-4757-a571-c79a14a9133f": {
    "category": "General Knowledge",
    "answers": ["Liver", "Skin", "Heart", "large Intestine"],
    "correctAnswer": "Skin",
    "id": "83cea789-1c6d-4757-a571-c79a14a9133f",
    "question": "What is the largest organ of the human body?",
    "difficulty": "easy"
  },
  "c051d919-2618-4f4d-bb42-5e13bdf58423": {
    "category": "General Knowledge",
    "answers": ["Vanilla", "Saffron", "Cinnamon", "Cardamom"],
    "correctAnswer": "Saffron",
    "id": "c051d919-2618-4f4d-bb42-5e13bdf58423",
    "question": "What is the world&#039;s most expensive spice by weight?",
    "difficulty": "medium"
  },
  "9289f1e9-f0c7-4fe4-a4f9-755745eb0bc8": {
    "category": "Entertainment: Video Games",
    "answers": ["Diamond/Pearl", "Ruby/Sapphire", "X/Y", "Black/White"],
    "correctAnswer": "Diamond/Pearl",
    "id": "9289f1e9-f0c7-4fe4-a4f9-755745eb0bc8",
    "question": "Which Pokemon generation did the fan-named &quot;Masuda Method&quot; first appear in? ",
    "difficulty": "easy"
  },
  "4db885c9-d4c1-4515-9e20-db71d6557cd6": {
    "category": "Entertainment: Video Games",
    "answers": [
      "Tateyuki Shigaraki",
      "Manfred Von Karma",
      "Gregory Edgeworth",
      "Bansai Ichiyanagi"
    ],
    "correctAnswer": "Bansai Ichiyanagi",
    "id": "4db885c9-d4c1-4515-9e20-db71d6557cd6",
    "question": "In the &quot;Ace Attorney&quot; series, who was the truly responsible of the forging of the autopsy report of the pivotal IS-7 incident?",
    "difficulty": "hard"
  },
}

// mock questions stats
export const mockDifficultyStats: TStat = {
  medium: 3,
  hard: 2,
  easy: 3,
  "all difficulties": 8
}

// mock category stats for different difficulties
export const mockCategoryStats: TStat = {
  "Entertainment: Film": 2,
  "Entertainment: Video Games": 2,
  "General Knowledge": 2,
  "Entertainment: Music": 2,
  "all categories": 8
}

export const mockEasyStats: TStat = {
  "Entertainment: Film": 1,
  "Entertainment: Video Games": 1,
  "General Knowledge": 1,
  "all categories": 3
}

export const mockHardStats: TStat = {
  "Entertainment: Music": 1,
  "Entertainment: Video Games": 1,
  "all categories": 2
}

export const mockMediumStats: TStat = {
  "Entertainment: Film": 1,
  "General Knowledge": 1,
  "Entertainment: Music": 1,
  "all categories": 3
}

// mock question
export const mockQuestion: IQuestion = {
  question: "Mock &lt;Question&gt;",
  id: "mockId",
  difficulty: "easy",
  correctAnswer: "Mock Correct Answer",
  category: "Mock Category",
  answers: [
    "Mocha &amp; Chai",
    "React &#124; Vue &#124; Angular",
    "Mock Aswer 3",
    "Mock Correct Answer",
  ],
};
