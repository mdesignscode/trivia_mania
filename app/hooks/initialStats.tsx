"use client";

import { GlobalContext } from "@/app/context/globalContext";
import { TStatsRequest } from "@/models/customRequests";
import { CATEGORIES } from "@/utils/localStorage_utils";
import { useQuery } from "@tanstack/react-query";
import axios from "axios";
import { useCallback, useContext, useEffect, useState } from "react";

type TStat = Record<string, number>;
export interface IInitialStats {
  difficultyStats: TStat;
  categoryStats: TStat;
  categoriesLoading: boolean;
  difficultiesLoading: boolean;
}

export default function useInitialStats(): IInitialStats {
  const [initialStats, setInitialStats] = useState<IInitialStats>({
    difficultyStats: {},
    categoryStats: {},
    categoriesLoading: true,
    difficultiesLoading: true,
  });
  const {
    storageIsAvailable,
    triviaUser,
    setCategoryChoice,
    playFilters: { difficulty }
  } = useContext(GlobalContext);

  const [shouldFetchCategories, setShouldFetchCategories] = useState(false);
  const [shouldFetchDifficulties, setShouldFetchDifficulties] = useState(false);

  // prepare url for question stats
  const baseUrl = process.env.NEXT_PUBLIC_API_BASE_URL;
  const url = baseUrl + "questions/stats";

  // fetch category stats
  const { data: categoryStats, isFetched: fetchedCategories } = useQuery<
    Record<string, number>
  >({
    queryKey: ["categoryStats", triviaUser],
    queryFn: async () => {
      try {
        const { data } = await axios.post(url, {
          recordType: "categories",
          difficulty: difficulty,
          userId: triviaUser?.id,
        } as TStatsRequest);
        return data;
      } catch (error) {
        console.log(error);
      }
    },
    initialData: {},
    enabled: shouldFetchCategories,
  });

  // fetch difficulty stats
  const { data: difficultyStats, isFetched: fetchedDifficulties } = useQuery<
    Record<string, number>
  >({
    queryKey: ["difficultyStats", triviaUser],
    queryFn: async () => {
      try {
        const { data } = await axios.post(url, {
          recordType: "difficulties",
          userId: triviaUser?.id,
        } as TStatsRequest);
        return data;
      } catch (error) {
        console.log(error);
      }
    },
    initialData: {},
    enabled: shouldFetchDifficulties,
  });

  const getStats = useCallback(() => {
    // fetch difficulties
    setShouldFetchDifficulties(true);
    if (fetchedDifficulties) {
      setInitialStats((state) => ({
        ...state,
        difficultyStats,
        difficultiesLoading: false,
      }));
      setShouldFetchDifficulties(false);
    }

    // fetch categories
    setShouldFetchCategories(true);
    if (fetchedCategories) {
      setInitialStats((state) => ({
        ...state,
        categoryStats,
        categoriesLoading: false,
      }));
      setShouldFetchCategories(false);
    }
  }, [categoryStats, difficultyStats, fetchedCategories, fetchedDifficulties]);

  // fetch stats
  useEffect(() => {
    getStats();
  }, [getStats]);

  // set previous categories if available
  useEffect(() => {
    if (storageIsAvailable && Object.keys(categoryStats).length) {
      const prevCategories = localStorage.getItem(CATEGORIES);

      if (prevCategories) {
        const preferences = prevCategories.split(",");

        if (preferences.length > 0) {
          const newState: Record<string, boolean> = {};
          Object.keys(categoryStats)
            .sort()
            .forEach((category) => {
              newState[category] = preferences.includes(category)
                ? true
                : false;
            });

          setCategoryChoice(Object.values(newState));
        }
      }
    }
  }, [categoryStats, setCategoryChoice, storageIsAvailable]);

  return initialStats;
}
