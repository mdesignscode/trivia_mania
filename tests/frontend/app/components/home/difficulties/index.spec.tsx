import Difficulties from "@/components/home/difficulties";
import { IInitialStats } from "@/hooks/initialStats";
import { mockDifficultyStats } from "@/utils/mockData";
import {
  mockCategoriesStats,
  renderHomeContext,
} from "@/utils/test_home_context";
import { screen } from "@/utils/test_utils";

// mock custom hook `useInititialStats`
jest.mock("@/hooks/initialStats", () => {
  return jest.fn(
    () =>
      ({
        difficultyStats: mockDifficultyStats,
        categoryStats: mockCategoriesStats,
        previousDifficulty: "",
      })
  );
});

describe("RenderDifficulties component", () => {
  it("should render the `RenderCategories` component", async () => {
    renderHomeContext(<Difficulties />, {
      currentUI: {
        welcome: false,
        difficulties: true,
        categories: false,
        play: false,
      },
    });
    const container = await screen.findByTestId(
      "render-difficulties-container"
    );
    expect(container).toBeInTheDocument();
  });

  it("snapshot matches", () => {
    // render component
    const { baseElement } = renderHomeContext(<Difficulties />, {currentUI: {
      welcome: false,
      difficulties: true,
      categories: false,
      play: false,
    }});
    expect(baseElement).toMatchSnapshot();
  });
});
