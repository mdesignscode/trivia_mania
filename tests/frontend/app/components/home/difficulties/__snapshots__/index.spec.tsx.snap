// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RenderDifficulties component snapshot matches 1`] = `
<body>
  <div>
    <div>
      <div>
        <div
          class="text-center col gap-3"
          data-testid="render-difficulties-container"
        >
          <h1
            class="text-xl"
          >
            Choose difficulty
          </h1>
          <div
            class="flex gap-2 justify-center flex-wrap"
          >
            <span
              data-testid="easy"
            >
              <button
                class="sc-bcPKhP fjXGcT"
              >
                easy
                 (
                3
                )
              </button>
            </span>
            <span
              data-testid="hard"
            >
              <button
                class="sc-bcPKhP fjXGcT"
              >
                hard
                 (
                2
                )
              </button>
            </span>
            <span
              data-testid="medium"
            >
              <button
                class="sc-bcPKhP fjXGcT"
              >
                medium
                 (
                3
                )
              </button>
            </span>
            <span
              data-testid="all difficulties"
            >
              <button
                class="sc-bcPKhP fjXGcT"
              >
                all difficulties
                 (
                8
                )
              </button>
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
